mission/TAC_tasks/scripts/docking/docking.py

```diff
@@ -0,0 +1,297 @@
+        max_error = 0.5
+        error = np.sqrt(
+            pow(
+                self.odom.pose.pose.position.x -
+                self.object.objectPose.pose.position.x, 2) + pow(
+                    self.odom.pose.pose.position.y -
+                    self.object.objectPose.pose.position.y, 2))
+        if (error < max_error):
+            return True
+        return False
+
+    def find_relative_to_mass_centre(self, offsetFromMC):
+        tf_quaternion = [
+            self.object.objectPose.pose.orientation.x,
+            self.object.objectPose.pose.orientation.y,
+            self.object.objectPose.pose.orientation.z,
+            self.object.objectPose.pose.orientation.w
+        ]
+
+        rotationMatrix = quaternion_matrix(tf_quaternion)
+        offsetArray = rotationMatrix.dot(offsetFromMC)
+
+        offsetPoint = Point()
+        offsetPoint.x = offsetArray[0]
+        offsetPoint.y = offsetArray[1]
+        offsetPoint.z = offsetArray[2]
+
+        return offsetPoint
+
+    def should_send_new_goal(self):
+        error = distance_between_points(self.object.objectPose.pose.position,
+                                        self.current_goal_pose.position)
+        a = 0.04
+        b = 0.02
+        error_limit = a * distance_between_points(
+            self.odom.pose.pose.position,
+            self.object.objectPose.pose.position) + b
+        if error > error_limit:
+            return True
+        return False
+
+    def execute(self, userdata):
+
+        self.state_pub.publish("docking/execute")
+
+        # Power puck relative to the center of mass (temporary until we get static transform)
+        powerPuckOffset = [0, 0, -0.6]
+
+        goal = dpGoal()
+        goal.x_ref = self.object.objectPose.pose
+
+        # Shifts DP goal from Docking_point to center of mass
+        offsetPoint = self.find_relative_to_mass_centre(powerPuckOffset)
+        goal.x_ref.position.x = goal.x_ref.position.x + offsetPoint.x
+        goal.x_ref.position.y = goal.x_ref.position.y + offsetPoint.y
+        goal.x_ref.position.z = self.odom.pose.pose.position.z
+
+        goal.DOF = [True, True, True, False, False, True]
+
+        self.dp_client.wait_for_server()
+        self.dp_client.send_goal(goal)
+        self.current_goal_pose = goal.x_ref
+
+        rate = rospy.Rate(10)
+        sending_rate = rospy.Rate(1)
+        sending_rate.sleep()
+
+        while (not rospy.is_shutdown() and self.isEnabled
+               and not self.reached_dp_goal):
+
+            self.object = self.landmarks_client("docking_point").object
+
+            goal.x_ref = self.object.objectPose.pose
+
+            offsetPoint = self.find_relative_to_mass_centre(powerPuckOffset)
+            goal.x_ref.position.x = goal.x_ref.position.x + offsetPoint.x
+            goal.x_ref.position.y = goal.x_ref.position.y + offsetPoint.y
+            goal.x_ref.position.z = goal.x_ref.position.z + offsetPoint.z
+
+            if not self.above_docking_point():
```

Comment:
Should be additional check for when dp should be turned on when pushing the drone down (from height < e.g. 2m and > e.g. 0.5 m) and when the dp should be turned off when pushing down (<0.5m to the docking point)